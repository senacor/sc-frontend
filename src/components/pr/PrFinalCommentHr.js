import React, { Component } from 'react';
import { connect } from 'react-redux';

import { withStyles } from '@material-ui/core/styles/index';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Grid from '@material-ui/core/Grid';
import * as actions from '../../actions';
import { getFinalCommentHr, getUserroles } from '../../reducers/selector';
import { debounce } from '../../helper/debounce';
import { translateContent } from '../translate/Translate';
import PrTextField from './PrTextField';
import TextFieldService from '../../service/TextFieldService';
import { injectIntl } from 'react-intl';

const styles = theme => ({
  bootstrapInput: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '3pt 3pt',
    '&:focus': {
      borderColor: '#80bdff',
      boxShadow: '0 0 0 0.2rem rgba(0,123,255,.25)'
    }
  },
  comment: {
    color: theme.palette.primary['400'],
    fontStyle: 'italic'
  }
});

class PrFinalCommentHr extends Component {
  constructor(props) {
    super(props);
    let { finalCommentHr } = this.props;
    let comment = finalCommentHr ? finalCommentHr : '';
    this.state = {
      commentText: comment
    };
  }

  handleChangeComment = prById => event => {
    this.setState({ commentText: event.target.value });

    this.sendComment(prById.id, event.target.value);
  };
  sendComment = debounce(this.props.changeFinalCommentHr, 500);

  render() {
    let {
      prById,
      finalCommentHr,
      readOnly,
      open,
      isActionPerformer,
      intl
    } = this.props;
    let { commentText } = this.state;

    let service = new TextFieldService();
    service.setIsActionPerformer(isActionPerformer);
    service.setReadOnlyFlag(readOnly);
    service.setOpenEditing(open);
    service.setReadOnlyText(finalCommentHr);
    service.setWriteableText(commentText);

    let helperText = intl.formatMessage({
      id: 'prfinalcommenthr.hronly'
    });

    return (
      <List>
        <ListItem>
          <Grid container direction={'column'}>
            <Grid item xs={12} sm={12} md={12} lg={12} xl={12}>
              <PrTextField
                fieldId={'FINAL_COMMENT_HR'}
                label={translateContent('FINAL_COMMENT_HR')}
                state={service.getState()}
                value={service.getValue()}
                helperText={helperText}
                onChange={this.handleChangeComment(prById)}
              />
            </Grid>
          </Grid>
        </ListItem>
      </List>
    );
  }
}

export const StyledComponent = withStyles(styles)(PrFinalCommentHr);
export default injectIntl(
  connect(
    state => ({
      finalCommentHr: getFinalCommentHr()(state),
      userroles: getUserroles(state)
    }),
    {
      changeFinalCommentHr: actions.changeFinalCommentHr
    }
  )(StyledComponent)
);
