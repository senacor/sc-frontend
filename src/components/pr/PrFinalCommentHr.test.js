import React from 'react';
import { StyledComponent } from './PrFinalCommentHr';
import { createShallow } from '@material-ui/core/test-utils';
import ROLES from '../../helper/roles';

describe('PrFinalCommentHr Component', () => {
  let shallow = createShallow({ dive: true });

  const prById = {
    statuses: [
      'ALL_DATES_ACCEPTED',
      'FILLED_SHEET_EMPLOYEE',
      'FILLED_SHEET_REVIEWER',
      'MODIFICATIONS_ACCEPTED_REVIEWER',
      'MODIFICATIONS_ACCEPTED_EMPLOYEE'
    ],
    finalCommentHr: 'I can type now.'
  };

  it('displays the PrFinalCommentHr', () => {
    const component = shallow(
      <StyledComponent
        prById={prById}
        userroles={[ROLES.PR_HR]}
        disabledText={false}
        open={true}
        readOnly={false}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('displays an info text before the employee has finalized', () => {
    let newStatuses = [
      'ALL_DATES_ACCEPTED',
      'FILLED_SHEET_EMPLOYEE',
      'FILLED_SHEET_REVIEWER',
      'MODIFICATIONS_ACCEPTED_REVIEWER'
    ];
    let newPr = Object.assign({}, prById, { statuses: newStatuses });
    const component = shallow(
      <StyledComponent
        prById={newPr}
        userroles={[ROLES.PR_HR]}
        disabledText={false}
        open={false}
        readOnly={false}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('displays the text read-only after HR has finalized', () => {
    let newStatuses = [
      'ALL_DATES_ACCEPTED',
      'FILLED_SHEET_EMPLOYEE',
      'FILLED_SHEET_REVIEWER',
      'MODIFICATIONS_ACCEPTED_REVIEWER',
      'MODIFICATIONS_ACCEPTED_EMPLOYEE',
      'PR_COMPLETED'
    ];
    let readOnlyPr = Object.assign({}, prById, { statuses: newStatuses });
    const component = shallow(
      <StyledComponent
        prById={readOnlyPr}
        userroles={[ROLES.PR_HR]}
        disabledText={false}
        open={true}
        readOnly={true}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
