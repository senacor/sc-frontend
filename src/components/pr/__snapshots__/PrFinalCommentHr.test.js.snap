// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrFinalCommentHr Component displays an info text before the employee has finalized 1`] = `
<WithStyles(List)>
  <WithStyles(ListItem)>
    <WithStyles(Grid)
      container={true}
      direction="column"
    >
      <WithStyles(Grid)
        item={true}
        lg={12}
        md={12}
        sm={12}
        xl={12}
        xs={12}
      >
        <WithStyles(Typography)
          component="p"
        >
          Abschließender Kommentar durch HR-Mitarbeiter (optional)
        </WithStyles(Typography)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        lg={12}
        md={12}
        sm={12}
        xl={12}
        xs={12}
      >
        <WithStyles(Typography)
          className="PrFinalCommentHr-comment-2"
          id="finalCommentHrReadonly"
          variant="body1"
        >
          Bitte nach Abschluss durch den Mitarbeiter ausfüllen.
        </WithStyles(Typography)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ListItem)>
</WithStyles(List)>
`;

exports[`PrFinalCommentHr Component displays the PrFinalCommentHr 1`] = `
<WithStyles(List)>
  <WithStyles(ListItem)>
    <WithStyles(Grid)
      container={true}
      direction="column"
    >
      <WithStyles(Grid)
        item={true}
        lg={12}
        md={12}
        sm={12}
        xl={12}
        xs={12}
      >
        <WithStyles(Typography)
          component="p"
        >
          Abschließender Kommentar durch HR-Mitarbeiter (optional)
        </WithStyles(Typography)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        lg={12}
        md={12}
        sm={12}
        xl={12}
        xs={12}
      >
        <TextField
          InputLabelProps={
            Object {
              "shrink": true,
            }
          }
          InputProps={
            Object {
              "classes": Object {
                "input": "PrFinalCommentHr-bootstrapInput-1",
              },
              "disableUnderline": true,
              "name": "comment",
            }
          }
          fullWidth={true}
          helperText="Möchte ich letzte Anmerkungen zum Performance Review machen?
                          Gibt es noch etwas zu ergänzen? Ich bestätige, dass ich das Review zur Kenntnis genommen habe."
          id="finalCommentHr"
          margin="none"
          multiline={true}
          onChange={[Function]}
          placeholder="Möchte ich letzte Anmerkungen zum Performance Review machen? Gibt es noch etwas zu ergänzen?"
          required={false}
          rows="4"
          rowsMax="10"
          select={false}
          value=""
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ListItem)>
</WithStyles(List)>
`;

exports[`PrFinalCommentHr Component displays the text read-only after HR has finalized 1`] = `
<WithStyles(List)>
  <WithStyles(ListItem)>
    <WithStyles(Grid)
      container={true}
      direction="column"
    >
      <WithStyles(Grid)
        item={true}
        lg={12}
        md={12}
        sm={12}
        xl={12}
        xs={12}
      >
        <WithStyles(Typography)
          component="p"
        >
          Abschließender Kommentar durch HR-Mitarbeiter (optional)
        </WithStyles(Typography)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        lg={12}
        md={12}
        sm={12}
        xl={12}
        xs={12}
      >
        <WithStyles(Typography)
          className="PrFinalCommentHr-comment-2"
          id="finalCommentHrReadonly"
          variant="body1"
        >
          Kein Eintrag.
        </WithStyles(Typography)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ListItem)>
</WithStyles(List)>
`;
